### DISCUSSION TITLE
TCA breaks source editor and autocomplete in Xcode 14

### DISCUSSION THREAD

**Discussion by tyler927** [pointfreeco/swift-composable-architecture#1264] [Q&A]
I cannot seem to use TCA in Xcode 14 (I've tried beta 4 and 5), which is odd to me because the latest Point-Free videos seem to be using Xcode 14 just fine. 

Once I import `ComposableArchitecture` into a file, pretty much everything I type seems to crash the editor and autocomplete showing the `An internal error occurred. Source editor functionality is limited. Attempting to restore..` message. 

I created a brand new SwiftUI iOS app in Xcode 14, add TCA as a dependency, make a new file, import TCA, and just typing `struct` crashes the editor. (This happens in my person apps, and the Case Studies project as well)

<img width="540" alt="Screen Shot 2022-08-13 at 9 36 19 AM" src="https://user-images.githubusercontent.com/10719676/184503172-dd64f747-3858-4184-8b1b-9e502b51c049.png">

Things I've tried: Clean build, delete derived data, reboot, make project in Xcode 13 and then open in 14, none of these seem to help. 

It does all work perfect in my Xcode 13.
I'm on a M1 Pro 16" with macOS 12.4.

Anyone else experiencing this, or have any ideas how to fix?

**Comment by mbrandonw** [comment]
Very strange, I'm not able to reproduce this:

<img width="436" alt="Screen Shot 2022-08-13 at 12 52 34 PM" src="https://user-images.githubusercontent.com/135203/184503508-76cdd0a5-7a1a-4f6b-9950-4db1e16e7ddf.png">

Just to confirm, are you pulling in the latest release of TCA? 0.39.0? What happens if you try pulling in some older versions? Does it ever get more stable?

**Reply by tyler927** [reply]
Yeah I was originally on 0.39.0, I just tried 0.38.0 and 0.35.0 and neither worked :/

**Reply by tgrapperon** [reply]
I confirm, I'm able to reproduce on my machine, Xcode 14b5 on Ventura beta 5, new vanilla multiplatform project, add TCA through Xcode packages, "import ComposableArchitecture" and then start to type `struct`.

<img width="470" alt="Xcode error message while performing the sequence of actions explained above" src="https://user-images.githubusercontent.com/35562418/184511481-45056cb2-5670-4fc8-b6c6-9cfdd44232dd.png">

I'm also experiencing this annoying message (it shifts the editor downward) very frequently on my SwiftPM project, but I didn't correlate with TCA. I've also checked with a new project that imports `swift-algorithms` instead of TCA and there is no issue there.

The error message doesn't explain anything obvious. I'll check the logs if I find something.

**Reply by mbrandonw** [reply]
Interesting! 

@tyler927, are you also on Ventura? I'm on Monterey.

**Reply by tyler927** [reply]
No @mbrandonw I am on Monterey as well, 12.4 to be exact. I also messed around with re-installing Xcode 14 in a different location, as well as clearing any Xcode and SPM cache I could find to no avail.

**Reply by tgrapperon** [reply]
[tcatest.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9332231/tcatest.zip)
Here is the archive of a project with the issue in any case.
The same project opened with Xcode 14b3 or b4 doesn't show any error message.
I've just filed FB11244375.

**Reply by tyler927** [reply]
Oh beta 3 does work for me too! Interestingly, beta 4 did not work for me though.

**Reply by tgrapperon** [reply]
Indeed, the error appears on beta 4 too. It takes a few seconds to appear for the first time, after Xcode finishes to start, and I didn't wait long enough when checking with different versions earlier.

**Reply by tgrapperon** [reply]
I was able to track down the source of the issue to likely be the declaration of the `SchedulerOf` typealias in `combine-schedulers`, or more precisely the declaration of a typealias of a protocol with primary associated type with the associated type specified on the rhs of the declaration.

For example, with a project depending only on `combine-scheduler`, the issue appears as soon as we add, in `AnyScheduler.swift`:
```swift
public typealias SchedulerOf<S: Scheduler> = Scheduler<S.SchedulerTimeType>
```
If one writes this declaration in the app instead, even in another file, there is no issue.
I'm able to reproduce with a simple SwiftPM package that only contains
``` swift
public protocol Foo<Bar> {
  associatedtype Bar
}
public typealias Problem = Foo<String>
```
If we import this package, the issue happens. If we declare instead:
```swift
public typealias Problem = Foo
```
there is no issue. So it seems triggered by using the primary associated value in an external module. The following is a minimal repro. Simply open `Package.swift`, and start to edit `EditThis.swift`, and the issue should appear.
[XcodeIssue.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9333421/XcodeIssue.zip)

I didn't audit all other dependencies of TCA (or TCA itself) for aliases to protocol with primary associated types. This is also fairly new so I can imagine a few rough edges. I've updated my FB with this information.

EDIT: I also opened this bug report in any case the problem comes from Swift: https://github.com/apple/swift/issues/60550

**Reply by stephencelis** [reply]
@tgrapperon Do you have a feedback to share that we can duplicate? This is an unfortunate bug that we would love to help escalate.

The good news is that protocol-style reducers we're experimenting with retain decent autocomplete, etc., even with Xcode in this broken state.

At the end of the day, we may have to hold off on some of these type aliases if they are problematic at the time of the Xcode going RC.

**Reply by tgrapperon** [reply]
@stephencelis [This gist](https://gist.github.com/tgrapperon/977f0406f67c417a4730a8640ab50526) is what corresponds to the latest state of the FB, as my initial submission was made when I still had not tracked down the source of the problem. I then updated the FB with this information.

Removing these aliases in local forks fixes the issue in my project, so it likely confirms that these typealiases are solely responsible for the error message in TCA projects (it could have been something else/more, as the error message is rather generic).

It builds fine even with the issue, so I suspect the problem to be purely on Xcode's side.  
Furthermore these typealiases are very recent, and likely not used broadly. The unique released occurrence I'm aware of is in `combine-schedulers` (and I don't think it could be deprecated without still causing the message). There are also one unused occurrence in the `protocol` branches (`ReducerProtocolOf`). It seems to affect only protocols with primary associated types, and other typealiases over generic types are fine fortunately.

**Reply by mbrandonw** [reply]
@tgrapperon @tyler927 Thanks for bringing this to our attention and for doing so much research to help us track it down.

We think it's probably best to remove the typealias for now. [Here's](https://github.com/pointfreeco/combine-schedulers/pull/64) a PR with some discussion on why the typealias isn't even useful today. Once we merge and release that you should be able to update to get a more stable editor.

However, though we have removed it from combine-schedulers, we do have future need for this pattern in the protocol features that are coming soon. We hope we can elevate the bugs with Apple to see if there's a possibility to fix it before we are ready to fully release the protocol stuff. 😐

**Reply by tgrapperon** [reply]
In any case, I've managed to cook up this declaration that doesn't cause warnings:
```swift
public protocol SchedulerOf<S>: Scheduler {
  associatedtype S: Scheduler where S.SchedulerTimeType == SchedulerTimeType
}
```
as something that approximately bears the same function/signature as:
```swift
public typealias SchedulerOf<S: Scheduler> = Scheduler<S.SchedulerTimeType>
```
The constraint over the generic is stronger however, so I doubt it can be used everywhere the typealias was able to be.

**Reply by mbrandonw** [reply]
@tgrapperon Interesting, thanks for investigating the alternatives.

We just released [0.7.2](https://github.com/pointfreeco/combine-schedulers/releases/tag/0.7.2) of combine schedulers which should fix the editor stability problems.

**Reply by tyler927** [reply]
Wow nice work tracking that down and getting a fix out for now @mbrandonw @tgrapperon , hopefully Apple can get that root issue actually fixed soon. I can confirm that 0.7.2 of combine schedulers has fixed my issues in beta 5 👍

**Reply by tgrapperon** [reply]
Confirmed fixed in Xcode 14.1!
https://github.com/apple/swift/pull/60801
https://github.com/apple/swift/issues/60550
